vault/core.go

```diff
@@ -916,6 +930,101 @@ func (c *Core) unsealPart(config *SealConfig, key []byte) ([]byte, error) {
 	return masterKey, nil
 }
 
+// Checks local version against data version. If data version doesn't
+// exist, it will attempt to write it. This must be called with the
+// state write lock held.
+func (c *Core) versionCheck() error {
+	// Get data version
+	rawDataVer, err := c.barrier.Get(coreDataVersionPath)
+	if err != nil {
+		c.logger.Error("core: data version check failed", "error", err)
+		c.barrier.Seal()
+		c.logger.Warn("core: vault is sealed")
+		return err
+	}
+	// If data version is empty, skip version comparison
+	if rawDataVer == nil {
+		// Handle case for secondary
+		if c.replicationState == consts.ReplicationSecondary {
+			c.logger.Warn("core: data version not found, ignoring version check")
+			c.versionCompatible = true
+			return nil
+		}
+		return c.persistVersion()
+	}
+
+	verStruct := &versionStruct{}
+	err = jsonutil.DecodeJSON(rawDataVer.Value, verStruct)
+	if err != nil {
+		c.logger.Warn("core: could not decode version information", "version_val", string(rawDataVer.Value))
+		return err
+	}
+	dataVer, err := gov.NewVersion(verStruct.Version)
+	if err != nil {
+		c.logger.Warn("core: could not perform data version check", "saved_version", verStruct.Version)
+		return err
+	}
+
+	// Get local version
+	localVer, err := gov.NewVersion(version.GetVersion().VersionNumber())
+	if err != nil {
+		// Don't do anything; this could be an unknown case such as during
+		// tests; we also don't want to persist the local version if we can't
+		// parse it
+		c.logger.Warn("core: could not perform data version check", "local_version", version.GetVersion().VersionNumber())
+		return err
+	}
+
+	// Perform comparison data
+	switch localVer.Compare(dataVer) {
+	case -1:
+		c.logger.Error("core: downgrade detected and not supported, re-sealing")
+		c.barrier.Seal()
+		c.logger.Warn("core: vault is sealed")
+		c.versionCompatible = false
+		return fmt.Errorf("downgrade detected and not supported; to proceed anyways delete %q and unseal Vault again", coreDataVersionPath)
+	case 1:
+		return c.persistVersion()
+	case 0:
+		c.versionCompatible = true
+	}
+
+	return nil
+}
+
+// persistVersion writes the local core version as the data version.
+// Data version is only updated by the primary.
+func (c *Core) persistVersion() error {
+	if c.replicationState == consts.ReplicationSecondary {
+		return fmt.Errorf("cannot write to %s by secondary", coreDataVersionPath)
```

Comment:
The secondary should never write but it shouldn't be an error to hit this path since you're not actually checking if it's a secondary before calling it. This should just return nil.